kong/clustering/rpc/socket.lua

```diff
@@ -0,0 +1,287 @@
+
+  return res
+end
+
+
+-- start reader and writer thread and event loop
+function _M:start()
+  self.read_thread = ngx.thread.spawn(function()
+    local last_seen = ngx_time()
+
+    while not exiting() do
+      local data, typ, err = self.wb:recv_frame()
+
+      if err then
+        if not is_timeout(err) then
+          return nil, err
+        end
+
+        local waited = ngx_time() - last_seen
+        if waited > CLUSTERING_PING_INTERVAL then
+          local res, err = self.outgoing:push(PING_TYPE)
+          if not res then
+            return nil, "unable to send ping: " .. err
+          end
+        end
+
+        if waited > PING_WAIT then
+          return nil, "did not receive ping frame from other end within " ..
+                      PING_WAIT .. " seconds"
+        end
+
+        -- timeout
+        goto continue
+      end
+
+      last_seen = ngx_time()
+
+      if typ == "ping" then
+        local res, err = self.outgoing:push(PONG_TYPE)
+        if not res then
+          return nil, "unable to handle ping: " .. err
+        end
+
+        goto continue
+      end
+
+      if typ == "pong" then
+        ngx_log(ngx_DEBUG, "[rpc] got PONG frame")
+
+        goto continue
+      end
+
+      if typ == "close" then
+        return true
+      end
+
+      assert(typ == "binary")
+
+      local payload = cjson_decode(data)
+      assert(payload.jsonrpc == "2.0")
+
+      if payload.method then
+        -- invoke
+
+        local cb = self.manager.callbacks.callbacks[payload.method]
+        if not cb then
+          local res, err = self.outgoing:push({
+            jsonrpc = "2.0",
+            id = payload.id,
+            ["error"] = {
+              code = jsonrpc.METHOD_NOT_FOUND,
+              message = "Method not found",
+            }
+          })
+          if not res then
+            return nil, "unable to handle ping: " .. err
+          end
+
+          goto continue
+        end
+
+        -- call dispatch
+
+        assert(ngx.timer.at(0, function(premature)
+          if premature then
+            return
+          end
+
+          local res, err = cb(self.node_id, unpack(payload.params))
+          if not res then
+            ngx_log(ngx_WARN, "[rpc] RPC callback failed: ", err)
+
+            res, err = self.outgoing:push({
+              jsonrpc = "2.0",
+              id = payload.id,
+              ["error"] = {
+                code = jsonrpc.SERVER_ERROR,
+                message = tostring(err),
+              }
+            })
+            if not res then
+              ngx_log(ngx_WARN, "[rpc] unable to push RPC call error: ", err)
+            end
+
+            return
+          end
+
+          -- success
+          res, err = self.outgoing:push({
+            jsonrpc = "2.0",
+            id = payload.id,
+            result = res,
+          })
+          if not res then
+            ngx_log(ngx_WARN, "[rpc] unable to push RPC call result: ", err)
+          end
+        end))
+
```

Comment:
Could we add a `goto continue` to remove `else` statement?

Reply:
Again, I am trying to avoid using jump labels as much as possible, to avoid abuses.